src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -6182,7 +6267,15 @@ private static bool DoMultipleAddressConnectCallback(object result, MultipleAddr
                 {
                     try
                     {
-                        context.socket.EndConnect((IAsyncResult)result);
+                        if (SocketPal.SupportsMultipleConnectAttempts || context.isUserConnectAttempt)
+                        {
+                            context.socket.EndConnect((IAsyncResult)result);
+                        }
+                        else
+                        {
+                            Debug.Assert(context.lastAttemptSocket != null);
+                            context.lastAttemptSocket.EndConnect((IAsyncResult)result);
+                        }
```

Comment:
I think that would only simplify this branch. That said, it may still be worth it--I'll take a closer look.